### PR TITLE
Fixes #26384 - cleaned os icons and new RH logo

### PR DESCRIPTION
**Pull request by lzap**
Big cleanup in our icon directory. Previously we had all icons in the
same folder as other images. This separates it, I created `osico`
subdirectories with two versions `16x16` and `64x64` in flat directory
structure. I wanted something short but "os" seemed generic, thus
"osico".

Previously we had icons in various sizes from some of which were not eve
square:

```
app/assets/images/AIX.png:                PNG image data, 21 x 21, 16-bit/color RGB, non-interlaced
app/assets/images/Altlinux.png:           PNG image data, 21 x 21, 16-bit/color RGB, non-interlaced
app/assets/images/Archlinux.png:          PNG image data, 24 x 24, 8-bit/color RGBA, non-interlaced
app/assets/images/bullet_toggle_plus.png: PNG image data, 16 x 16, 8-bit grayscale, non-interlaced
app/assets/images/Centos.png:             PNG image data, 21 x 20, 8-bit/color RGBA, non-interlaced
app/assets/images/Coreos.png:             PNG image data, 21 x 21, 8-bit/color RGBA, non-interlaced
app/assets/images/Darwin.png:             PNG image data, 21 x 21, 16-bit/color RGB, non-interlaced
app/assets/images/Debian.png:             PNG image data, 21 x 25, 8-bit colormap, non-interlaced
app/assets/images/edit-hover.png:         PNG image data, 16 x 16, 8-bit colormap, non-interlaced
app/assets/images/edit.png:               PNG image data, 16 x 16, 4-bit colormap, non-interlaced
app/assets/images/Fedora.png:             PNG image data, 21 x 21, 8-bit/color RGBA, non-interlaced
app/assets/images/Freebsd.png:            PNG image data, 21 x 21, 8-bit/color RGBA, non-interlaced
app/assets/images/Gentoo.png:             PNG image data, 21 x 21, 8-bit/color RGBA, non-interlaced
app/assets/images/Junos.png:              PNG image data, 21 x 21, 16-bit/color RGB, non-interlaced
app/assets/images/login_logo.png:         PNG image data, 266 x 117, 8-bit/color RGBA, non-interlaced
app/assets/images/navbar.png:             PNG image data, 1400 x 60, 8-bit/color RGBA, non-interlaced
app/assets/images/NXOS.png:               PNG image data, 21 x 21, 8-bit colormap, non-interlaced
app/assets/images/OracleLinux.png:        PNG image data, 21 x 21, 16-bit/color RGB, non-interlaced
app/assets/images/pie_chart_icon.png:     PNG image data, 17 x 17, 8-bit/color RGBA, non-interlaced
app/assets/images/pie_overlay.png:        PNG image data, 117 x 125, 8-bit gray+alpha, non-interlaced
app/assets/images/Puppet.png:             PNG image data, 21 x 21, 16-bit/color RGB, non-interlaced
app/assets/images/Rancheros.png:          PNG image data, 21 x 21, 8-bit/color RGBA, non-interlaced
app/assets/images/Redhat.png:             PNG image data, 21 x 20, 8-bit/color RGBA, non-interlaced
app/assets/images/Scientific.png:         PNG image data, 24 x 24, 8-bit/color RGBA, non-interlaced
app/assets/images/SLC.png:                PNG image data, 24 x 25, 8-bit/color RGBA, non-interlaced
app/assets/images/Solaris.png:            PNG image data, 21 x 21, 16-bit/color RGB, non-interlaced
app/assets/images/sort_asc_disabled.png:  PNG image data, 19 x 19, 8-bit/color RGBA, non-interlaced
app/assets/images/sort_asc.png:           PNG image data, 19 x 19, 8-bit/color RGBA, non-interlaced
app/assets/images/sort_both.png:          PNG image data, 19 x 19, 8-bit gray+alpha, non-interlaced
app/assets/images/sort_desc_disabled.png: PNG image data, 19 x 19, 8-bit gray+alpha, non-interlaced
app/assets/images/sort_desc.png:          PNG image data, 19 x 19, 8-bit/color RGBA, non-interlaced
app/assets/images/Suse.png:               PNG image data, 21 x 21, 8-bit/color RGBA, non-interlaced
app/assets/images/Ubuntu.png:             PNG image data, 21 x 21, 8-bit/color RGBA, non-interlaced
app/assets/images/VRP.png:                PNG image data, 21 x 21, 8-bit/color RGBA, non-interlaced
app/assets/images/Windows.png:            PNG image data, 21 x 21, 16-bit/color RGB, non-interlaced
app/assets/images/Xenserver.png:          PNG image data, 21 x 21, 16-bit/color RGB, non-interlaced
```

All images were redownloaded and resampled from scratch from
high-quality versions and I also created copies of higher-quality size
64x64 for future purpose. We might want to show the icon in better
quality in Host detail page. Let's insist on both qualities now during
reviews anytime someone adds new icon.

```
[lzap@box foreman]$ file app/assets/images/osico16x16/*
app/assets/images/osico16x16/archlinux.png:  PNG image data, 16 x 16, 8-bit/color RGBA, non-interlaced
app/assets/images/osico16x16/centos.png:     PNG image data, 16 x 16, 8-bit/color RGBA, non-interlaced
app/assets/images/osico16x16/coreos.png:     PNG image data, 16 x 16, 8-bit/color RGBA, non-interlaced
app/assets/images/osico16x16/debian.png:     PNG image data, 16 x 16, 8-bit/color RGBA, non-interlaced
app/assets/images/osico16x16/fedora.png:     PNG image data, 16 x 16, 8-bit/color RGBA, non-interlaced
app/assets/images/osico16x16/freebsd.png:    PNG image data, 16 x 16, 8-bit/color RGBA, non-interlaced
app/assets/images/osico16x16/gentoo.png:     PNG image data, 16 x 16, 8-bit/color RGBA, non-interlaced
app/assets/images/osico16x16/rancher.png:    PNG image data, 16 x 16, 8-bit/color RGBA, non-interlaced
app/assets/images/osico16x16/redhat.png:     PNG image data, 16 x 16, 8-bit/color RGBA, non-interlaced
app/assets/images/osico16x16/scientific.png: PNG image data, 16 x 16, 8-bit/color RGBA, non-interlaced
app/assets/images/osico16x16/stub:           directory
app/assets/images/osico16x16/suse.png:       PNG image data, 16 x 16, 8-bit/color RGBA, non-interlaced
app/assets/images/osico16x16/ubuntu.png:     PNG image data, 16 x 16, 8-bit/color RGBA, non-interlaced
[lzap@box foreman]$ file app/assets/images/osico64x64/*
app/assets/images/osico64x64/archlinux.png:  PNG image data, 64 x 64, 8-bit/color RGBA, non-interlaced
app/assets/images/osico64x64/centos.png:     PNG image data, 64 x 64, 8-bit/color RGBA, non-interlaced
app/assets/images/osico64x64/coreos.png:     PNG image data, 64 x 64, 8-bit/color RGBA, non-interlaced
app/assets/images/osico64x64/debian.png:     PNG image data, 64 x 64, 8-bit/color RGBA, non-interlaced
app/assets/images/osico64x64/fedora.png:     PNG image data, 64 x 64, 8-bit/color RGBA, non-interlaced
app/assets/images/osico64x64/freebsd.png:    PNG image data, 64 x 64, 8-bit/color RGBA, non-interlaced
app/assets/images/osico64x64/gentoo.png:     PNG image data, 64 x 64, 8-bit/color RGBA, non-interlaced
app/assets/images/osico64x64/rancher.png:    PNG image data, 64 x 64, 8-bit/color RGBA, non-interlaced
app/assets/images/osico64x64/redhat.png:     PNG image data, 64 x 64, 8-bit/color RGBA, non-interlaced
app/assets/images/osico64x64/scientific.png: PNG image data, 64 x 64, 8-bit/color RGBA, non-interlaced
app/assets/images/osico64x64/stub:           directory
app/assets/images/osico64x64/suse.png:       PNG image data, 64 x 64, 8-bit/color RGBA, non-interlaced
app/assets/images/osico64x64/ubuntu.png:     PNG image data, 64 x 64, 8-bit/color RGBA, non-interlaced
```

There is a subdirectory called `stub` which holds all "stub" images
created via https://github.com/lzap/foreman-dummy-icons script which we
use for icons with unapproved use or unknown licenses. Previously we
copied those files randomly but now it's much more clear what is a stub
and what is not thank to the same naming scheme.

All filenames are now lower case explicitly. This was also creating some confusion
in the past (and bugs).

All images were filtered with `pngcrush` tool saving approximately 40%
of the original (GIMP saved) size.

Edit: The patch now contains also the new RH logo.

### CODE REVIEW

File: app/helpers/operatingsystems_helper.rb [hunk 1]
```diff
@@ -8,52 +8,57 @@ def label(record)
...
+             "firebrick-o"
+           when /CoreOS|ContainerLinux|Container Linux/i
+             "coreos"
+           when /RancherOS/i
+             "rancheros"
+           when /NXOS/i
+             "stub/darkslateblue-n"
+           when /XenServer/i
+             "stub/black-x"
+           when /Puppet/i
```
Comment (xprazak2) [inline app/helpers/operatingsystems_helper.rb:54]: Is Puppet an operating system family?

**Comment by xprazak2** [comment]
Nice cleanup, though visuals might need some additional styling:

icons

Also, no icons appear for RancherOS, Xen and couple of others.

Test failures relevant, asset pipeline seems to ignore newly added files.

**Comment by timogoebel** [comment]
The icons don't look transparent in the screenshot above.

File: app/helpers/operatingsystems_helper.rb (hunk 1 above)
Comment (lzap) [reply app/helpers/operatingsystems_helper.rb:54]: I have no idea why this is there. The icon was present however, probably to keep it as I don't want to change behavior in this PR?

**Comment by lzap** [comment]
Fixed the nbsp and also the dummy icons are now rectangle with round corners rather than circle which was challenging to align to center on 16x16 canvas. Transparency problem was also fixed. The patch was updated to contain the new logo:

image

With high-quality 64x64 versions we could also put them to host detail page, unsure where to. I would like to do this in a separate patch probably.

Edit: Ignore centos vs redhat inconsistency, I have my OSes defined incorrectly.

**Comment by timogoebel** [comment]
@lzap: We need to keep the puppet icon in the original place so we don't break https://github.com/theforeman/foreman/blob/d3ee92b2fb87e5add467027693158b6446d6847f/app/helpers/fact_values_helper.rb#L48-L51. This is also used by plugins.

**Comment by lzap** [comment]
Well I have deleted it because I could not find license of the puppet logo. I have replaced it with a dummy icon. Therefore I will put a dummy icon there for plugins, unless someone else is lucky to find or get an approval.

**Comment by lzap** [comment]
Maybe @ohadlevy can decide? We've been shipping it for a decade... Let's just keep it?

**Comment by timogoebel** [comment]
I believe we can remove it as an Operatingsystem. The Fact Icon is already a dummy icon, see https://github.com/theforeman/foreman/pull/3874#issuecomment-248848656.

**Comment by lzap** [comment]
Oh ok, no licensing issues then. Well, I'd prefer to fix plugins rather than keeping 'Puppet.png' in the codebase forever. Any ideas what exactly do I need to send PRs into?

**Comment by timogoebel** [comment]
> Any ideas what exactly do I need to send PRs into?

That'll force plugins to maintain a new branch. Imho that's a lot of effort for little gain. I believe you might be able to just change the path for Puppet here:
https://github.com/theforeman/foreman/blob/5579da3f4072c9f937a7340e4c850414f726fa95/app/models/fact_names/puppet_fact_name.rb#L2-L4

If that does not work and you do have to update any plugins, every config management plugin needs an update:

* foreman_ansible
* foreman_chef
* foreman_omaha
* foreman_salt

**Comment by lzap** [comment]
I think I misunderstood you, now I get it. Well, how about simply refactoring the helper method in a way that it works? I've also renamed the directory from `osico` to just `icons` because it can eventually hold any kind of icon, including facter stubs.

```ruby
  # Return fact icon image tag. Stub is used in some cases because
  # of legal requirements. Make sure to get legal advice prior
  # putting any logos into our git repository.
  def fact_origin_icon(origin, opts = {})
    return origin if origin == 'N/A'
    size = opts[:size] ||= '16x16'
    filename = (origin =~ /Puppet/) ? "stub/goldenrod-p" : origin
    image_tag("icons#{size}/#{filename}.png", title: origin, size: size)
  end
```

**Comment by timogoebel** [comment]
> Well, how about simply refactoring the helper method in a way that it works?

You would still need to move the asset files in all the plugins. Also I don't like the hardcoded value for Puppet. We have put quite of effort into making the architecture pluggable. If you want to change the icon path, please do that in https://github.com/theforeman/foreman/blob/5579da3f4072c9f937a7340e4c850414f726fa95/app/models/fact_names/puppet_fact_name.rb#L2-L4.

**Comment by lzap** [comment]
> If you want to change the icon path, please do that in

This is IMHO worse, plugins actually do this: `if xyz.origin == "Puppet"` I think it is OpenSCAP doing this.

> Also I don't like the hardcoded value for Puppet

It follows what we do in OS, and OS is also "pluggable" or "open" for users to define anything they want.

I believe this is sane approach, if the string contains "puppet" replace it. Once we extract puppet code into a plugin then we can simply provide any path we want and remove this. However if you insist, I can change this and fix OpenSCAP. However semantically the method will not make sense: "give me origin", "ok it's stub/goldenrod-p".

**Comment by timogoebel** [comment]
> It follows what we do in OS

This is where I don't follow.

**Comment by lzap** [comment]
> It follows what we do in OS

What I am trying to say is that the helper is essentially similar to what we do in OS - case with regular expressions. Well, it is much more simple case.

I don't like `origin` to return `stub/goldenrod-p` string that's nasty. If you want a different solution please suggest something cleaner.

**Comment by lzap** [comment]
Is there anything I can do to move forward with this?

**Comment by netsandbox** [comment]
@lzap the small Red Hat image has the wrong size:

```
file app/assets/images/icons16x16/redhat.png 
app/assets/images/icons16x16/redhat.png: PNG image data, 64 x 64, 8-bit/color RGBA, non-interlaced
```

**Comment by timogoebel** [comment]
> Is there anything I can do to move forward with this?

I need to take a closer look myself. If anybody wants to jump in in the meantime: Go ahead.

**Comment by lzap** [comment]
Rebased, I would really need to get this in, @xprazak2 or @ezr-ondrej or @ares ?

File: app/helpers/fact_values_helper.rb [hunk 2]
```diff
@@ -45,13 +45,15 @@ def show_full_fact_value(fact_value)
...
+  def fact_origin_icon(origin, opts = {})
     return origin if origin == 'N/A'
-    title = origin
-    if origin.include? "/"
-      title = origin.split("/")[1]
-    end
-    image_tag(origin + ".png", :title => title)
+    origin = origin.split("/")[1] if origin.include?("/")
+    size = opts[:size] ||= '16x16'
+    filename = (origin =~ /Puppet/) ? "stub/goldenrod-p" : origin
```
Comment (timogoebel) [inline app/helpers/fact_values_helper.rb]: Note that we define the icon for plugins here.
https://github.com/theforeman/foreman_ansible/blob/f5b602a76c925f147f3ff74194b496a38c1cdc69/app/models/foreman_ansible/fact_name.rb#L8-L10

Puppet specific code should go here:
https://github.com/theforeman/foreman/blob/develop/app/models/fact_names/puppet_fact_name.rb

Comment (lzap) [reply app/helpers/fact_values_helper.rb]: Hmmm I cannot use origin, it's defined (and documented) as origin name, e.g. Puppet or Ansible. Ansible plugin uses that in a weird way, let me go step back and define this properly - I am going to fix `origin` to be used properly and introduce new method `icon_path` to return what's really needed. Plugins can redefine, sane default will be provided.

**Comment by lzap** [comment]
How about this? Now I am going to file PRs to katello, ansible and issue warning on Discourse for other plugin authors.

File: app/helpers/fact_values_helper.rb (hunk 2 above)
Comment (ares) [reply app/helpers/fact_values_helper.rb]: I think the point here is, you replaced the logic how we determine icon names for facts from various soruces. Ansible, Chef, Salt and Katello relies on this. Unless you implement `icon_path` method for their parsers, they will end up with `icons16x16/stub/black-x.png`.

Comment (ares) [reply app/helpers/fact_values_helper.rb]: which is now resolved I think by sending PRs to other repos

**Comment by timogoebel** [comment]
> How about this?

I like it.

[test upgrade]
[test foreman]
[test katello]

**Comment by lzap** [comment]
I like that you like it! All green.

**Comment by timogoebel** [comment]
> All green.

It skips all ci tests. I do think we want to run them.

**Comment by ezr-ondrej** [comment]
> It skips all ci tests. I do think we want to run them.

If we do, we might not be green :thinking:
[test upgrade]

Edit: it is stubborn and keeps the first information given ... no tests

**Comment by timogoebel** [comment]
[test upgrade]
[test foreman]
[test katello]

**Comment by timogoebel** [comment]
I removed the skip ci comments in the PR description. Tests are now running.

**Comment by lzap** [comment]
"DEPRECATION WARNING: The asset "icons16x16/firebrick-h.png" is not present in the asset pipeline.Falling back to an asset that may be in the public folder." - what it wants me to do?

File: app/helpers/operatingsystems_helper.rb [hunk 3]
```diff
@@ -6,54 +6,62 @@ def label(record)
...
+           when /SLC/i
+             "stub/blue-s"
+           when /FreeBSD/i
+             "freebsd"
+           when /aix/i
+             "stub/forestgreen-a"
+           when /Junos/i
+             "stub/darkblue-j"
+           when /VRP/i
+             "firebrick-h"
```
Comment (ezr-ondrej) [inline app/helpers/operatingsystems_helper.rb]: should be `stub/firebrick-h` I believe

File: app/helpers/operatingsystems_helper.rb [hunk 4]
```diff
@@ -6,54 +6,62 @@ def label(record)
...
+           when /OracleLinux/i
+             "firebrick-o"
```
Comment (ezr-ondrej) [inline app/helpers/operatingsystems_helper.rb]: ^^

**Review by ezr-ondrej — COMMENTED** [review]
I believe you need to prefix the `firebricks` with folder where they actually are.
Also what are the `firebrick-<letter>` do we need all of them?

**Comment by lzap** [comment]
Rebasing. Previously we were adding them when needed but people were lazy finding them, I think it's not big deal shipping few images I pngcrushed them, in total it's:

[lzap@box foreman]$ du -h app/assets/images/icons64x64/stub app/assets/images/icons16x16/stub/
1,3M    app/assets/images/icons64x64/stub
1,3M    app/assets/images/icons16x16/stub/

**Review by ezr-ondrej — APPROVED** [review]
LGTM :+1:

**Comment by ezr-ondrej** [comment]
@timogoebel, @xprazak2 any comments?

**Comment by xprazak2** [comment]
I have no additional comments.

**Comment by lzap** [comment]
@ezr-ondrej or @xprazak2 can you do the honor please? It's one of my sprint items :-)

Then we need to merge the three related plugin PRs too as this will break their tests.

**Comment by ezr-ondrej** [comment]
Thanks @lzap :+1: Merging!

**Comment by ezr-ondrej** [comment]
[theforeman/foreman_chef](https://github.com/theforeman/foreman_chef) is looking with pain in eyes :persevere: at @lzap ... or you meant we should push theforeman/foreman_chef#90 to match this PR?

**Pull request merged**